cli/tools/test.rs

```diff
@@ -967,6 +1079,164 @@ impl TestReporter for PrettyTestReporter {
   }
 }
 
+struct JunitTestReporter {
+  path: String,
+  // Stores TestCases (i.e. Tests) by the Test ID
+  cases: IndexMap<usize, quick_junit::TestCase>,
+}
+
+impl JunitTestReporter {
+  fn new(path: String) -> Self {
+    Self {
+      path,
+      cases: IndexMap::new(),
+    }
+  }
+
+  fn convert_status(status: &TestResult) -> quick_junit::TestCaseStatus {
+    match status {
+      TestResult::Ok => quick_junit::TestCaseStatus::success(),
+      TestResult::Ignored => quick_junit::TestCaseStatus::skipped(),
+      TestResult::Failed(failure) => quick_junit::TestCaseStatus::NonSuccess {
+        kind: quick_junit::NonSuccessKind::Failure,
+        message: Some(failure.to_string()),
+        ty: None,
+        description: None,
+        reruns: vec![],
+      },
+      TestResult::Cancelled => quick_junit::TestCaseStatus::NonSuccess {
+        kind: quick_junit::NonSuccessKind::Error,
+        message: Some("Cancelled".to_string()),
+        ty: None,
+        description: None,
+        reruns: vec![],
+      },
+    }
+  }
+
+  fn flush_report(
+    &self,
+    elapsed: &Duration,
+    tests: &IndexMap<usize, TestDescription>,
+  ) {
+    let mut suites: IndexMap<String, quick_junit::TestSuite> = IndexMap::new();
+    for (id, case) in &self.cases {
+      if let Some(test) = tests.get(id) {
+        suites
+          .entry(test.location.file_name.clone())
+          .and_modify(|s| {
+            s.add_test_case(case.clone());
+          })
+          .or_insert_with(|| {
+            quick_junit::TestSuite::new(test.location.file_name.clone())
+              .add_test_case(case.clone())
+              .to_owned()
+          });
+      }
+    }
+
+    let mut report = quick_junit::Report::new("deno test");
+    report.set_time(*elapsed).add_test_suites(
+      suites
+        .values()
+        .cloned()
+        .collect::<Vec<quick_junit::TestSuite>>(),
+    );
+
+    // How should we handle errors when writing out the report if at all?
+    if let Ok(mut file) = std::fs::File::create(self.path.clone()) {
+      let _ = file.write_all(report.to_string().unwrap().as_bytes());
+    }
```

Comment:
How should we handle errors when writing out the report (if at all)?

Reply:
Maybe I should promote `flush_report` to be a part of the trait and have it return `Result` 🤔 `PrettyTestReporter` would have an implementation that just returns `Ok` but other `TestReporters` could have real implementations

Reply:
Implemented in e7c409c